app/core/src/main/java/com/fsck/k9/Preferences.java

```diff
@@ -45,7 +46,8 @@ public static synchronized Preferences getPreferences(Context context) {
     private Map<String, Account> accounts = null;
     private List<Account> accountsInOrder = null;
     private Account newAccount;
-    private Context context;
+    private final ArrayList<AccountsChangeListener> accountsChangeListeners = new ArrayList<>();
```

Comment:
An alternative would be using [`CopyOnWriteArrayList`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CopyOnWriteArrayList.html) that is thread-safe. It is especially well suited for listeners (as the list of listeners changes rarely but is iterated frequently; the javadocs go into more detail on that matter).

Reply:
Nice. But all methods that trigger read access to this list are already synchronized anyway. The only place that is missing is adding and removing listeners.

Reply:
> But all methods that trigger read access to this list are already synchronized anyway.

That may also be problematic. For example `notifyListeners` invokes callbacks when the lock is held.

An excerpt from [Effective Java 2nd Edition](https://www.amazon.com/Effective-Java-2nd-Joshua-Bloch/dp/0321356683/), Item 67:

> To avoid liveness and safety failures, never cede control to the client within a synchronized method or block. In other words, inside a synchronized region, do not invoke a method that is designed to be overridden, or *one provided by a client in the form of a function object* [emphasis mine] (Item 21). From the perspective of the class with the synchronized region, such methods are alien. The class has no knowledge of what the method does and has no control over it. Depending on what an alien method does, calling it from a synchronized region can cause exceptions, deadlocks, or data corruption.

Item 67 has also a nice example of what could go wrong (but I'd like to avoid pasting the entire chapter here :) ).

Reply:
not sure if CopyOnWriteArrayList is the perfect candidate here, since the list of listeners changes all the time (e.g. onPause/onResume of activities), but accounts don't change *that* often. Anyways we're gonna have the copy on write or on read, so I just went with it.

Reply:
> Anyways we're gonna have the copy on write or on read, so I just went with it.

Looks legit :+1:, also :+1:  for `GuardedBy` :)